packages/devtools_app/lib/src/charts/treemap.dart

```diff
@@ -0,0 +1,493 @@
+import 'package:flutter/material.dart';
+
+import '../trees.dart';
+
+class Treemap extends StatefulWidget {
+  const Treemap({
+    @required this.rootNode,
+    @required this.levelsVisible,
+    @required this.width,
+    @required this.height,
+    @required this.onTap,
+  });
+
+  final TreemapNode rootNode;
+
+  // The depth of children visible from this treemap widget.
+  // For example, levelsVisible = 2
+  // ---------------
+  // |    l = 1    |
+  // |  ---------  |
+  // |  | l = 2 |  |
+  // |  |       |  |
+  // |  |       |  |
+  // |  ---------  |
+  // ---------------
+  final int levelsVisible;
+
+  final double width;
+  final double height;
+
+  final VoidCallback onTap;
+
+  @override
+  _TreemapState createState() => _TreemapState();
+}
+
+enum PivotType { pivotByMiddle, pivotBySize }
+
+class _TreemapState extends State<Treemap> {
+  PivotType pivotType = PivotType.pivotBySize;
+  static const double minHeightToDisplayText = 20.0;
+  bool shouldDisplayText;
+
+  TreemapNode rootNode;
+
+  @override
+  void initState() {
+    super.initState();
+    rootNode = widget.rootNode;
+    shouldDisplayText = widget.height > minHeightToDisplayText;
+  }
+
+  void cellOnTap(String name) {
+    final TreemapNode child = rootNode.childrenMap[name];
+    if (child != null) {
+      setState(() {
+        rootNode = child;
+      });
+    }
+  }
+
+  // Computes the total size of a given list of treemap nodes.
+  int computeSize(List<TreemapNode> children, int start, int end) {
+    int sum = 0;
+    for (int i = start; i <= end; i++) {
+      sum += children[i].byteSize;
+    }
+    return sum;
+  }
+
+  int computePivot2(List<TreemapNode> children) {
+    switch (pivotType) {
+      case PivotType.pivotByMiddle:
+        {
+          return (children.length / 2).floor();
+        }
+      case PivotType.pivotBySize:
+        {
+          int pivotIndex = -1;
+          double maxSize = double.negativeInfinity;
+          for (int i = 0; i < children.length; i++) {
+            if (children[i].byteSize > maxSize) {
+              maxSize = children[i].byteSize.toDouble();
+              pivotIndex = i;
+            }
+          }
+          return pivotIndex;
+        }
+    }
+    return -1;
+  }
+
+  // Divides a given list of treemap nodes into four parts:
+  // L1, P, L2, L3
+
+  // P (pivot) is the treemap node chosen to be the pivot based on the pivot type.
+
+  // L1 inclues all treemap nodes before the pivot treemap nodes.
```

Comment:
typo: includes
typo: nodes --> node at the end of the sentence

Reply:
Fixed